### ISSUE TITLE
Accessing previous rows (LAG/LEAD - window function style)

### ISSUE THREAD

**Issue by svladykin** [ClickHouse/ClickHouse#5485]
As far as I see currently there is no way to access previous rows like `lag` window function in SQL? https://www.postgresql.org/docs/11/functions-window.html

Still, the function `runningAccumulate` accesses the values from the previous rows (only within a single Block??).
Looks like it must be relatively easy to implement function `runningLag` the same way?

**Comment by den-crane** [comment]
```
SELECT  t.1 AS k,t.2 AS v
FROM(SELECT arrayJoin([('a', 1), ('a', 2), ('b', 1), ('a', 3), ('b', 2), ('b', 100)]) AS t)

┌─k─┬───v─┐
│ a │   1 │
│ a │   2 │
│ b │   1 │
│ a │   3 │
│ b │   2 │
│ b │ 100 │
└───┴─────┘


SELECT
    k,
    arrayDifference(groupArray(v))
FROM
(
    SELECT
        t.1 AS k,
        t.2 AS v
    FROM
    (
        SELECT arrayJoin([('a', 1), ('a', 2), ('b', 1), ('a', 3), ('b', 2), ('b', 100)]) AS t
    )
)
GROUP BY k

┌─k─┬─arrayDifference(groupArray(v))─┐
│ b │ [0,1,98]                       │
│ a │ [0,1,1]                        │
└───┴────────────────────────────────┘
```

**Comment by svladykin** [comment]
@den-crane 
Thanks, but this does not answer any of my questions :)
Of course it is possible to `groupArray`/`arrayJoin` back and forth, but it is not always convenient and probably it will not be very efficient for a huge number of rows..

**Comment by den-crane** [comment]
OK. The only problem where is no function which support (windows/partition by). 
runningAccumulate calculates totalSum without resetting on a next group.

if need this then your runningLag is equal runningDifference

SELECT k, runningDifference(v) 
FROM (SELECT intDiv(number, 2) AS k, number v FROM numbers(10) ORDER BY k,v);

**Comment by svladykin** [comment]
@den-crane  Thanks for the explanation, this makes total sense.
So, if I have a table with web page visits like `visits(visitor, date)` and want to know how many total unique users and new unique users came each date then it should look like the following?
```sql
select date, total_uniq_visitors, runningDifferenceStartingWithFirstValue(total_uniq_visitors) this_date_uniq_visitors
from (
select date, runningAccumulate(u) total_uniq_visitors
from (
select date, uniqState(visitor) u from visits 
group by date
order by date
))
```

**Comment by filimonov** [comment]
Please follow also main window function related issue #1469
There are also several workarounds listed there

**Comment by vincenzon** [comment]
I am hitting an issue using runningDifference that may be me misunderstanding how the function is applied, but here is an example riffing on the example from @den-crane . 

Here I difference, but there is a jump as the key k changes from 'a' to 'b'. 

```
SELECT
    t.1 AS k,
    t.2 AS v,
    runningDifference(v) AS d,
    runningDifference(xxHash32(t.1)) AS i
FROM
(
    SELECT arrayJoin([('a', 1), ('a', 2), ('a', 3), ('b', 11), ('b', 13), ('b', 15)]) AS t
)

┌─k─┬──v─┬─d─┬──────────i─┐
│ a │  1 │ 0 │          0 │
│ a │  2 │ 1 │          0 │
│ a │  3 │ 1 │          0 │
│ b │ 11 │ 8 │ 1291794793 │
│ b │ 13 │ 2 │          0 │
│ b │ 15 │ 2 │          0 │
└───┴────┴───┴────────────┘
```

So I inserted the hash value of the key thinking I could filter to the cases where the hash was equal to zero. But that doesn't work as it seems the runningDifference is lazily evaluated:

```
SELECT
    k,
    v,
    d,
    i
FROM
(
    SELECT
        t.1 AS k,
        t.2 AS v,
        runningDifference(v) AS d,
        runningDifference(xxHash32(t.1)) AS i
    FROM
    (
        SELECT arrayJoin([('a', 1), ('a', 2), ('a', 3), ('b', 11), ('b', 13), ('b', 15)]) AS t
    )
)
WHERE i = 0

┌─k─┬──v─┬──d─┬─i─┐
│ a │  1 │  0 │ 0 │
│ a │  2 │  1 │ 0 │
│ a │  3 │  1 │ 0 │
│ b │ 13 │ 10 │ 0 │
│ b │ 15 │  2 │ 0 │
└───┴────┴────┴───┘
```

I was expecting the fourth row to have a value of 2 for v.  Is this a bug or expected?

**Comment by den-crane** [comment]
@vincenzon enable_optimize_predicate_expression bug

19.4.3.11
```
┌─k─┬──v─┬─d─┬─i─┐
│ a │  1 │ 0 │ 0 │
│ a │  2 │ 1 │ 0 │
│ a │  3 │ 1 │ 0 │
│ b │ 13 │ 2 │ 0 │
│ b │ 15 │ 2 │ 0 │
└───┴────┴───┴───┘
```
19.10.1.669
```
┌─k─┬──v─┬──d─┬─i─┐
│ a │  1 │  0 │ 0 │
│ a │  2 │  1 │ 0 │
│ a │  3 │  1 │ 0 │
│ b │ 13 │ 10 │ 0 │
│ b │ 15 │  2 │ 0 │
└───┴────┴────┴───┘
SET enable_optimize_predicate_expression = 0
┌─k─┬──v─┬─d─┬─i─┐
│ a │  1 │ 0 │ 0 │
│ a │  2 │ 1 │ 0 │
│ a │  3 │ 1 │ 0 │
│ b │ 13 │ 2 │ 0 │
│ b │ 15 │ 2 │ 0 │
└───┴────┴───┴───┘
```

https://github.com/yandex/ClickHouse/issues/3885

**Comment by vincenzon** [comment]
Thanks @den-crane.

**Comment by leoones** [comment]
@den-crane  @vincenzon 
I want to subtract the values specified in the array，
I want to find the value of 1-2 and 1-3， 
┌─k─v---
│ a │  1 │ 
│ a │  2 │
│ a │  3 │

**Comment by filimonov** [comment]
[neighbor](https://clickhouse.tech/docs/en/query_language/functions/other_functions/#neighbor
) function introduced in #5925 is avaliable since 19.14 address that issue.

But it works currently inside each block separately:

```
SELECT 
    number, 
    neighbor(toNullable(number), 1)
FROM numbers(12)
ORDER BY number ASC
SETTINGS max_block_size = 4
FORMAT PrettyCompactMonoBlock

┌─number─┬─neighbor(toNullable(number), 1)─┐
│      0 │                               1 │
│      1 │                               2 │
│      2 │                               3 │
│      3 │                            ᴺᵁᴸᴸ │
│      4 │                               5 │
│      5 │                               6 │
│      6 │                               7 │
│      7 │                            ᴺᵁᴸᴸ │
│      8 │                               9 │
│      9 │                              10 │
│     10 │                              11 │
│     11 │                            ᴺᵁᴸᴸ │
└────────┴─────────────────────────────────┘

12 rows in set. Elapsed: 0.006 sec. 
```

**Comment by leoones** [comment]
> [neighbor](https://clickhouse.tech/docs/en/query_language/functions/other_functions/#neighbor) function introduced in #5925 is avaliable since 19.14 address that issue.
> 
> But it works currently inside each block separately:
> 
> ```
> SELECT 
>     number, 
>     neighbor(toNullable(number), 1)
> FROM numbers(12)
> ORDER BY number ASC
> SETTINGS max_block_size = 4
> FORMAT PrettyCompactMonoBlock
> 
> ┌─number─┬─neighbor(toNullable(number), 1)─┐
> │      0 │                               1 │
> │      1 │                               2 │
> │      2 │                               3 │
> │      3 │                            ᴺᵁᴸᴸ │
> │      4 │                               5 │
> │      5 │                               6 │
> │      6 │                               7 │
> │      7 │                            ᴺᵁᴸᴸ │
> │      8 │                               9 │
> │      9 │                              10 │
> │     10 │                              11 │
> │     11 │                            ᴺᵁᴸᴸ │
> └────────┴─────────────────────────────────┘
> 
> 12 rows in set. Elapsed: 0.006 sec. 
> ```

SELECT 
    a, 
    neighbor(toNullable(a), 1)
FROM 
(
    SELECT 
        arrayJoin([1, 2, 3]) AS a, 
        'u' AS b
)

┌─a─┬─neighbor(toNullable(a), 1)─┐
│ 1 │                          2 │
│ 2 │                          3 │
│ 3 │                       ᴺᵁᴸᴸ │
└───┴────────────────────────────┘

**The result I want is**   
select  1 - 2 ;   result -1
select  1 - 3 ;   result -2

The final display is as follows

┌─a─┬─neighbor(toNullable(a), 1)─┐
│ 1 │                          -1│
│ 2 │                          -2 │
│ 3 │                       ᴺᵁᴸᴸ │
└───┴────────────────────────────┘

**Comment by blinkov** [comment]
@svladykin, do you have any further questions?

**Comment by kiwimg** [comment]
when the window function will be supported in Clickhouse?
I hope to see the function early!

**Comment by alexey-milovidov** [comment]
@kiwimg The lag and lead functions are implemented here: #21895.
